Color support

johnnyr26 opened the pull request:
Adds checkbox that allows/disables color backgrounds
Generates new color if the default colors run out

johnnyr26 committed:
Implements the color option into the schedule

When the color is selected, the user would be able to get a color-coded schedule
In the case (shouldn't be but you never know with some schedules) that all of the colors are used up, the program would generate a new color with a light opacity in order for the class to be readable.

johnnyr26 committed:
Color background checkbox

Adds checkbox that will activate the color backgrounds in the schedule

johnnyr26 committed:
Adds color support

johnnyr26 committed:
Merge pull request #1 from JRProgramming/patch-2

Color background checkbox

johnnyr26 committed:
Fixes the implementation of the colors

johnnyr26 commented:
Fix anything you need to fix but I tried my best to make all changes necessary for the color implementation. I also tried my best to keep the style of the code as similar as it can be to yours.
I haven't tested it yet so there may be some mistakes I'm not aware of.

johnnyr26 committed:
Update HTMLExporter.js

johnnyr26 committed:
Update main.js

jbis9051 commented:
Why close?

johnnyr26 commented:
My bad, my mistake

johnnyr26 commented:
I tried to make a change to add semicolons to the lines I forgot to add semicolons, so I thought I needed to close it but I realized I didn't

jbis9051 commented on SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:
```diff
@@ -7,6 +7,7 @@ module.exports = (timedScheduleObject) => {
         $('table thead tr').append(`<th class="letter">${day}</th>`);
     });
     const numOfClasses = Object.values(timedScheduleObject)[0].length;
+    const colors = ["tomato", "dodgerblue", "yellow", "orange", "mediumseagreen", "lightblue", "pink", "#d6ff99", "gainsboro", "#FAE7B5", "#a89de1", "burlywood"]
```
Should be globally declared (same array each time)

johnnyr26 replied:
Okay

jbis9051 commented on SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:
```diff
@@ -34,6 +35,18 @@ module.exports = (timedScheduleObject) => {
                     }
+                    if (theClass.color) {
+                        if(colors[i]) {
+                            document.getElementsByClassName("class").style.backgroundColor = colors
```
DOM API not available within Node.js

johnnyr26 replied:
Look at the new file change that occurred after this

jbis9051 commented on SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:
```diff
@@ -34,6 +35,18 @@ module.exports = (timedScheduleObject) => {
+                                color.push(Math.floor(Math.random() * 255))
+                            }
+                            document.getElementsByClassName("class").style.backgroundColor = "rgba(" + color[0] + "," + color[1] + "," + color[2] + ", .5)"
+                            // In the case that all of the colors are used, the program would generate a new color with a light opacity in order for the class to be readable
```
We know there is a max amount of classes possible (6 classes a day times 8 days = 48 colors). I suggest we create a bank of 48 colors.

johnnyr26 replied:
Hard to do when the colors get really similar

jbis9051 replied:
You really only need like 10-12 solid colors, after that in can be same base color just a different shade as these will rarely ever come up.

jbis9051 commented on views/index.ejs:
```diff
@@ -71,6 +71,7 @@
     <div class="checkbox-wrapper"><label><input type="checkbox" checked id="teacher-toggle"><span>Show Teachers</span></label></div>
     <div class="checkbox-wrapper"><label><input type="checkbox" checked id="room-toggle"><span>Show Room</span></label></div>
     <div class="checkbox-wrapper"><label><input type="checkbox" checked id="free-toggle"><span>Show Free Periods</span></label></div>
+    <div class="checkbox-wrapper"><labeL><input type="checkbox" id="color-toggle"><span>Show Color Background</span></label></div>
```
Are we not showing the colors by default?

johnnyr26 replied:
We can do that

jbis9051 commented on SchedulerCreator/src/js/exporters/HTML/HTMLExporter.js:
```diff
@@ -27,7 +28,21 @@ module.exports = (timedScheduleObject) => {
+                                color.push(Math.floor(Math.random() * 255));
+                            }
+                            const newColor = `rgba("${colorArray[0]}","${colorArray[1]}", "${colorArray[2]}", .5)`;
+                            classCode = `<td class="class" style="background-color: ${newColor}">`;
```
How are we gonna disable this with the toggle?

jbis9051 commented on public/js/main.js:
```diff
@@ -142,6 +142,10 @@ $('#submit-button').addEventListener('click', e => {
                 $('#free-toggle').checked = false;
             }
+            if($$('table .color').length == 0) {
+                $('#color-toggle').setAttribute('disabled.','');
```
Typo: Extra "." for the "disabled" attribute

jbis9051 commented on public/js/main.js:
```diff
@@ -213,6 +217,14 @@ $('#free-toggle').addEventListener('change', evt => {
+    if($('#color-toggle').checked){
+        $$('table .color').forEach(el => el.removeAttribute('hidden'));
+    } else {
+        $$('table .color').forEach(el => el.setAttribute('hidden',''));
```
I don't think there is any elements that will match this selector.

jbis9051 reviewed (changes requested):
Please fix changes mentioned. Also I suggest editing the timedScheduleObject creation script to include indices.

johnnyr26 committed:
Color array is a global variable

jbis9051 commented:
> I suggest editing the timedScheduleObject creation script to include indices.

This has been resolved with https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/90c620c04d4848c9dea9d60d348a302bc426f91a. Use the `classNum` to index the color bank.

johnnyr26 committed:
Makes the color background default

johnnyr26 committed:
Changes index

johnnyr26 committed:
Update main.js

jbis9051 commented:
I suggest creating an array of 48 css names, instead of an array of color hexes/names. Then creating a CSS style sheet with: 
```CSS
.color .some_class_name {
     background: #some_color;
}
```
for each.  We can then simply toggle the `.color` class. This will allow us to toggle colors without having to keep track of the individual colors. 

Also for the colors themselves, I suggest soft colors that will show well with text.  For example, instead of `blue` maybe `#99badd`

johnnyr26 committed:
Update main.js

johnnyr26 commented:
I got work to do so I’m done for the day. I’ll talk and work tomorrow

Just one quick help. Don’t know JQuery for this, but can you write me the JQ code that will set the background color back to white

$('#color-toggle').addEventListener('change', evt => {
    if(!$('#color-toggle').checked){
        //Resets the background color to white
    }
});

Thanks.

> On Sep 25, 2019, at 11:07 PM, Josh Brown <notifications@github.com> wrote:
> 
> I suggest creating an array of 48 css names, instead of an array of color hexes/names. Then creating a CSS style sheet with:
> 
> .color .some_class_name {
>      background: #some_color;
> }
> for each. We can then simply toggle the .color class. This will allow us to toggle colors without having to keep track of the individual colors.
> 
> Also for the colors themselves, I suggest soft colors that will show well with text
>

jbis9051 commented:
This is not jQuery, see [L1](https://github.com/jbis9051/Blind-Brook-Calendar-Creator/blob/master/public/js/main.js#L1).

In jQuery it would be:

```javascript
$('#color-toggle').change(evt => {
    if(!$('#color-toggle').checked){ // however dup selector
       $(selector).css('background','white');
    }
});
```
Plain js:
```javascript
document.querySelector('#color-toggle').addEventListener('change', evt => {
    if(!document.querySelector('#color-toggle').checked){ 
       document.querySelectorAll(selector).forEach(el => el.style.background = "white");
    }
});
```

johnnyr26 committed:
Changes background color of table

Changes the background color of table back to white when the checkbox isn't checked

jbis9051 commented:
Please implement the updated color generation system (See: https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/e2ee45d9c8daa6bec2c9b5dc68fe94b50444f6a8). Also please separate https://github.com/jbis9051/Blind-Brook-Calendar-Creator/pull/3/commits/5b9020881ca13824354acbfb083d26817b8ee8c5. I will merge that.

jbis9051 commented:
I suggest getting an IDE. My suggestion is [Jet Brain's WebStorm](https://www.jetbrains.com/webstorm/). You can get a free student account.

johnnyr26 commented:
Can I get a list of all the things you want me to do?
It's hard to see what I have to do when they are all separated

jbis9051 commented:
1. Create a separate PR for https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/5b9020881ca13824354acbfb083d26817b8ee8c5
2. Create a new PR for Color support
3. Using the stuff from https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/90c620c04d4848c9dea9d60d348a302bc426f91a and https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/e2ee45d9c8daa6bec2c9b5dc68fe94b50444f6a8, make the colors and toggle work
**Don't update this PR, create a new one**

Optional but suggested:

1. Get an IDE (Webstorm)

jbis9051 commented:
Color support added with: https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/f264d17d7453af2901d6c1853593768422537952

johnnyr26 commented:
Hey,

I saw you were able to add the color support feature.

I’ll redirect my website to yours and tell all the administrators that are relying on my site once you add my name to the bottom of the site as well as yours like we agreed to.

Let me know if you need anything else.

> On Sep 28, 2019, at 12:28 PM, Josh Brown <notifications@github.com> wrote:
> 
> Color support added with: f264d17 <https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/f264d17d7453af2901d6c1853593768422537952>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/jbis9051/Blind-Brook-Calendar-Creator/pull/3?email_source=notifications&email_token=AG7B2LUZMHI2WRQLS2HEBIDQL6A4PA5CNFSM4I2UDCNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD725JDY#issuecomment-536204431>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AG7B2LVN6YDXETYMRWZL3L3QL6A4PANCNFSM4I2UDCNA>.
>

jbis9051 commented:
@JRProgramming Please see https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/441c723c743ca062d3f2e243ea48eeac1925b8f4. Please close the "Combine" issue on your project once you have redirected your site.

johnnyr26 commented:
My site now redirects to yours.

I’ll let them know when I see them that my url now redirects to your site and to use your url from now.

Glad we were able to combine

> On Sep 29, 2019, at 8:49 PM, Josh Brown <notifications@github.com> wrote:
> 
> @JRProgramming <https://github.com/JRProgramming> Please see 441c723 <https://github.com/jbis9051/Blind-Brook-Calendar-Creator/commit/441c723c743ca062d3f2e243ea48eeac1925b8f4>. Please close the "Combine" issue on your project once you have redirected your site.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/jbis9051/Blind-Brook-Calendar-Creator/pull/3?email_source=notifications&email_token=AG7B2LWPMQMWLI5TDJ3QCKDQMFELBA5CNFSM4I2UDCNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD74DGUY#issuecomment-536359763>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AG7B2LVJZG4CM6WOORN7MXTQMFELBANCNFSM4I2UDCNA>.
>
